dhall/src/Dhall/ETags.hs

```diff
@@ -0,0 +1,256 @@
+    } deriving (Eq, Ord)
+
+newtype Tags = Tags (M.Map FilePath [(PosInFile, Tag)])
+
+instance Semigroup Tags where
+    (Tags ts1) <> (Tags ts2) = Tags (M.unionWith (<>) ts1 ts2)
+
+instance Monoid Tags where
+    mempty = Tags M.empty
+    mappend = (<>)
+
+{-| For example, for line: `let foo = "foo"` tag is:
+    `Tag "let " "foo"`
+-}
+data Tag = Tag
+    { tagPattern :: Text
+      -- ^ In vtags source code this field has name pattern and EMacs use it as regex pattern
+      --   to locate line with tag. It's looking for ^<tag pattern>.
+      --   Looks like vi is not using it.
+    , tagName :: Text
+      -- ^ text, that editor compare with selected text. So it's really name of entity
+    }
+
+{-| Generate ETags for Dhall expressions
+-}
+generate
+    :: Input
+    -- ^ Path to directory or filename that should be indexed
+    -> [Text]
+    -- ^ List of suffixes for dhall files [".dhall"] by default
+    -> Bool
+    -- ^ Flag if generate should follow symlinks
+    -> IO Text
+    -- ^ Content for tags file
+generate inp sxs followSyms = do
+    isD <- case inp of
+               StandardInput -> return False
+               InputFile p -> doesDirectoryExist p
+    files <- inputToFiles followSyms (map T.unpack (if isD then sxs else [""])) inp
+    tags <- foldM (\ts f -> handle (\(SomeException _) -> return ts)
+                                  ((<>) ts . fileTags f <$> TIO.readFile f)) mempty files
+    return (showTags tags)
+
+{-| Find tags in Text (second argument) and generates list of them
+    To make tags for filenames working in both, emacs and vi, add two initial tags.
+    First for `filename` for vi and second with `/filename` for emacs.
+    Other tags are working for both.
+-}
+fileTags :: FilePath -> Text -> Tags
+fileTags f t = Tags (M.singleton f
+                    (initialMap <> getTagsFromText t))
+    where initialViTag = (FP 1 0, Tag "" (T.pack . takeFileName $ f))
+          initialEmacsTag = (FP 1 1, Tag "" ("/" <> (T.pack . takeFileName) f))
+          initialMap = [initialViTag, initialEmacsTag]
+
+getTagsFromText :: Text -> [(PosInFile, Tag)]
+getTagsFromText t = case eitherTags of
+    Right x -> x
+    _ -> mempty
+    where eitherTags = fixPosAndDefinition t . getTagsFromExpr <$> exprFromText "" t
+
+-- after getTagsFromExpr line and column in line are in PosInFile for each tag.
+-- And tagPattern is not added.
+-- Emacs use tag pattern to check if tag is on line. It compares line from start
+-- with tag pattern and in case they are the same, relocate user.
+-- fixPosAndDefinition change position to line and byte offset and add tag pattern.
+-- For example, for:
+-- ----------------
+-- let foo = "bar"
+-- let baz = "qux"
+-- ----------------
+-- Input for this function is:
+-- [(FP 0 4, "foo"), (FP 1 4, "baz")]
+-- And it will be transformed into:
+-- [(FP 0 4, Tag "let foo " "foo"), (FP 1 20, Tag "let baz " "baz")]
+-- where 20 is byte offset from file start.
```

Comment:
Great example! Thanks!

Please eventually put this into a proper doctest, like this one:

https://github.com/dhall-lang/dhall-haskell/blob/3e5cb3492222a1e10de07896ed9689df0a94dbd0/dhall/src/Dhall/Pretty/Internal.hs#L140-L149